### ISSUE TITLE
[Bug?] failure in copying multiple files from local to remote file system

### ISSUE THREAD

**Issue by kumare3** [fsspec/filesystem_spec#724]
According to the documentation of `filesystem.put`, a way to copy files is as follows
`f.put("/tmp/test/*", "s3://my-s3-bucket/blah2/", recursive=True)`
But the following line,
https://github.com/intake/filesystem_spec/blob/master/fsspec/spec.py#L194, which is invoked from
https://github.com/intake/filesystem_spec/blob/master/fsspec/spec.py#L797, will render the rpath value to be `s3://my-s3-bucket/blah`, which will not allow multiple files to be copied, but instead write one file with name `blah2`.

With this, we cannot use fsspec's put recursive for writing multiple files or copying over a directory. But, I replicated the internal function (to expand local paths and then create other_paths) to achieve a similar non async write.

**Comment by martindurant** [comment]
Can you please specify the actual method you called, and the state of the local files you are copying from and the state of the remote files before and after? Perhaps you can come up with a reproducible test showing the failure. Doing a little testing right now suggests to me that hings are working as expected.

**Comment by kumare3** [comment]
> Can you please specify the actual method you called, and the state of the local files you are copying from and the state of the remote files before and after? Perhaps you can come up with a reproducible test showing the failure. Doing a little testing right now suggests to me that hings are working as expected.

I was testing with s3. Let me see if I can reproduce with local and will share

**Comment by wild-endeavor** [comment]
Trying with the local file system and running into I think the same or related weirdness.  The behavior of put seems to change depending on whether or not the folder exists.  Can you recommend a way around this please?

For example, if I have
```
$ l -R ~/temp/data/source
total 8
-rw-r--r--  1 user  staff    12B Feb 24 14:32 original.txt
drwxr-xr-x  3 user  staff    96B Feb 24 15:42 nested/
-rw-r--r--  1 user  staff     0B Feb 24 15:42 empty.md
drwxr-xr-x  5 user  staff   160B Feb 24 15:42 ./
drwxr-xr-x  6 user  staff   192B Feb 24 15:51 ../

/Users/user/temp/data/source/nested:
total 8
drwxr-xr-x  3 user  staff    96B Feb 24 15:42 ./
-rw-r--r--  1 user  staff    11B Feb 24 15:42 more.txt
drwxr-xr-x  5 user  staff   160B Feb 24 15:42 ../
```

and I run

```python
def test_local_only_issue():
    source_dir = "/Users/user/temp/data/source/"

    local = fsspec.filesystem(None)
    new_temp_dir = tempfile.mkdtemp()
    print(f"Exists temp dir {new_temp_dir}")
    local.put(source_dir, new_temp_dir, recursive=True)

    new_temp_dir_2 = tempfile.mkdtemp()
    new_temp_dir_2 = os.path.join(new_temp_dir_2, "doesnotexist")
    local.put(source_dir, new_temp_dir_2, recursive=True)
    print(f"Did not exist temp dir {new_temp_dir_2}")
```

I get
```
$ l '/var/folders/jx/54tww2ls58n8qtlp9k31nbd80000gp/T/tmpl38k4sks/doesnotexist'
total 8
drwx------  3 user  staff    96B Feb 24 17:46 ../
-rw-r--r--  1 user  staff     0B Feb 24 17:46 empty.md
drwxr-xr-x  3 user  staff    96B Feb 24 17:46 nested/
drwxr-xr-x  5 user  staff   160B Feb 24 17:46 ./
-rw-r--r--  1 user  staff    12B Feb 24 17:46 original.txt
$ l '/var/folders/jx/54tww2ls58n8qtlp9k31nbd80000gp/T/tmpf7p76t__'
total 0
drwx------     3 user  staff    96B Feb 24 17:46 ./
drwxr-xr-x     5 user  staff   160B Feb 24 17:46 source/
drwx------@ 1181 user  staff    37K Feb 24 17:46 ../
```

I would like to get the top behavior, where it's just the files that get copied in.  Could you maybe give me a code snippet that accomplishes this?  I can do an exists check... but I actually want to just overwrite conflicting files.  Is there an overwrite arg I can use?

Thank you.

**Comment by martindurant** [comment]
@wild-endeavor : @ianthomas23 has been working on consistency and setting expectation for this kind of operation ( e.g., https://github.com/fsspec/filesystem_spec/pull/1148 ). What exact version are you using?

**Comment by wild-endeavor** [comment]
These are the versions I have now but happy to change!
```
$ pip show fsspec s3fs
Name: fsspec
Version: 2022.11.0
Summary: File-system specification
Home-page: http://github.com/fsspec/filesystem_spec
Author:
Author-email:
License: BSD
Location: /Users/ytong/envs/flytekit/lib/python3.10/site-packages
Requires:
Required-by: s3fs
---
Name: s3fs
Version: 2022.11.0
Summary: Convenient Filesystem interface over S3
Home-page: http://github.com/fsspec/s3fs/
Author:
Author-email:
License: BSD
Location: /Users/ytong/envs/flytekit/lib/python3.10/site-packages
Requires: aiobotocore, aiohttp, fsspec
Required-by:
```

**Comment by ianthomas23** [comment]
@wild-endeavor Historically there have been some inconsistencies in this area, but since PRs #1148 and #1163 these have been fixed apart from a few corner cases. Those PRs haven't made their way into a release yet but I expect that to happen in early March. The latest commit (`c75410b880`) on `master` branch running the following test code (based on yours but explicitly creating the directories and files to copy):

```python
import fsspec
import os
import tempfile

source_dir = "source/"
local = fsspec.filesystem('file')

# Set up source directory for test.
if local.exists(source_dir):
    local.rm(source_dir, recursive=True)
assert not local.exists(source_dir)
nested_dir = os.path.join(source_dir, "nested")
local.mkdir(nested_dir)
local.touch(os.path.join(source_dir, 'original.txt'))
local.touch(os.path.join(nested_dir, 'more.txt'))
print("source directory")
for i, f in enumerate(local.find(source_dir)):
    print('  ', i, f)

new_temp_dir = tempfile.mkdtemp()
print(f"Exists temp dir {new_temp_dir}")
local.put(source_dir, new_temp_dir, recursive=True)
for i, f in enumerate(local.find(new_temp_dir)):
    print('  ', i, f)

new_temp_dir_2 = tempfile.mkdtemp()
new_temp_dir_2 = os.path.join(new_temp_dir_2, "doesnotexist")
local.put(source_dir, new_temp_dir_2, recursive=True)
print(f"Did not exist temp dir {new_temp_dir_2}")
for i, f in enumerate(local.find(new_temp_dir_2)):
    print('  ', i, f)
```
gives
```
source directory
   0 /Users/iant/whatever/source/nested/more.txt
   1 /Users/iant/whatever/source/original.txt
Exists temp dir /var/folders/s5/7bphl1y95377dmj86tv2bwlm0000gp/T/tmpl627qegg
   0 /var/folders/s5/7bphl1y95377dmj86tv2bwlm0000gp/T/tmpl627qegg/nested/more.txt
   1 /var/folders/s5/7bphl1y95377dmj86tv2bwlm0000gp/T/tmpl627qegg/original.txt
Did not exist temp dir /var/folders/s5/7bphl1y95377dmj86tv2bwlm0000gp/T/tmpu7lowg0l/doesnotexist
   0 /var/folders/s5/7bphl1y95377dmj86tv2bwlm0000gp/T/tmpu7lowg0l/doesnotexist/nested/more.txt
   1 /var/folders/s5/7bphl1y95377dmj86tv2bwlm0000gp/T/tmpu7lowg0l/doesnotexist/original.txt
```
which I believe is what you want. Please correct me if I have misunderstood.

Note that under certain circumstances the behaviour is indeed different depending on whether the target directory exists or not. This is intentional and correct, it follows what command-line `cp` does. I am in the process of documentation all the various `cp`, `get` and `put` possibilities to make all of this clearer.

**Comment by wild-endeavor** [comment]
I think this works.  I see the current behavior if I drop the trailing `/` in the source path but with it there, I'm able to reproduce the behavior I'm looking for with that commit.  This will be included in the Feb monthend release?  Any idea when that will be going out?

Thank you.

**Comment by martindurant** [comment]
> when that will be going out?

we don't have a specific timeline, so "when I have some free time" :)

**Comment by wild-endeavor** [comment]
Actually, another oddity.  Now working with get.  Don't have the complete test code but basically

```
local.get("/tmp/empty_exists_1/", "/tmp/empty_exists_2/", recursive=True)
```
isn't doing what I would expect.  With the slashes there, I would expect no change, but it appears that `"/tmp/empty_exists_2/empty_exists_1"` is getting created.  Is there a way around this?

Thank you.

**Comment by wild-endeavor** [comment]
Yeah the behavior does seem to change based on whether the source directory is empty.  If it's not empty, the underlying files are copied over correctly, but if empty, it does appear to copy the parent name.

```python
local = fsspec.filesystem("file")

@pytest.fixture
def source_folder():
    # Set up source directory for testing
    parent_temp = tempfile.mkdtemp()
    src_dir = os.path.join(parent_temp, "source", "")
    nested_dir = os.path.join(src_dir, "nested")
    local.mkdir(nested_dir)
    local.touch(os.path.join(src_dir, "original.txt"))
    local.touch(os.path.join(nested_dir, "more.txt"))
    yield src_dir
    shutil.rmtree(parent_temp)


@pytest.fixture
def empty_source_folder():
    # Set up source directory for testing
    parent_temp = tempfile.mkdtemp()
    src_dir = os.path.join(parent_temp, "source", "")
    local.mkdir(src_dir)
    yield src_dir
    shutil.rmtree(parent_temp)

def test_local_provider_get_l(source_folder):
    assert source_folder.endswith("/")
    with tempfile.TemporaryDirectory() as dest_tmpdir:
        dest_tmpdir = dest_tmpdir + "/"
        local.get(source_folder, dest_tmpdir, recursive=True)

        print(f"\nDest non empty {dest_tmpdir}")
        for i, f in enumerate(local.find(dest_tmpdir, withdirs=True)):
            print('  ', i, f)


def test_local_provider_get_empty_l(empty_source_folder):
    assert empty_source_folder.endswith("/")
    with tempfile.TemporaryDirectory() as dest_tmpdir:
        dest_tmpdir = dest_tmpdir + "/"
        local.get(empty_source_folder, dest_tmpdir, recursive=True)

        print(f"\nDest copying empty {dest_tmpdir}")
        for i, f in enumerate(local.find(dest_tmpdir, withdirs=True)):
            print('  ', i, f)
```

**Comment by ianthomas23** [comment]
Yes, I have confirmed that `cp`, `get` and `put` of empty directory isn't correct. @wild-endeavor Can you create a new issue for this please.

**Comment by martindurant** [comment]
@wild-endeavor just wondering: do you really have a use case for get/put with a local filesystem (which should be identical to cp), or is this just a proxy for what you really want to do?

**Comment by wild-endeavor** [comment]
I can work around this yes, but also was just testing.  The first behavior that's already been fixed on master affected s3 as well.  I'd also like to keep the code on my end as clean as possible - it's much easier to always pass a string to fsspec, get a filesystem, and call get/put on it.

In the course of normal operation of what i'm working on, we frequently switch between local, s3, and gcs.

I will cut a new issue for this.  Thank you both.
